### PR TITLE
Hw18

### PR DESCRIPTION
**Pull request by Olga-Bagrova**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,49 @@
...
+* count_percentage - return percentage of each amino acid in sequence
+
+**Filtration fastq-files**:
+* `filter_fastq` - filter fastq-sequences based on its gc-content, length and quality.
+
+**telegram_logger**:
+* `@telegram_logger` - decorator for logging launches of declared functions. Upon completion, it sends messages through the telegram bot.
+ 
+**API** 
+* `run_genscan` - python API for [the service](http://hollywood.mit.edu/GENSCAN.html).
```
Comment (iam28th) [inline README.md:32]: я бы в тексте указал GENSCAN на случай если людям будет лень ссылку проверять)

File: README.md [hunk 2]
```diff
@@ -1 +1,49 @@
...
+* transcribe - return transcribed sequence
+
+**Protein only**:
+* count_percentage - return percentage of each amino acid in sequence
+
+**Filtration fastq-files**:
+* `filter_fastq` - filter fastq-sequences based on its gc-content, length and quality.
+
+**telegram_logger**:
+* `@telegram_logger` - decorator for logging launches of declared functions. Upon completion, it sends messages through the telegram bot.
```
Comment (iam28th) [inline README.md:29]: 1) В случае перечисления имён из кода в readme придётся вручную следить за их синхронизацией, если решите что-то где-то переименовать; возможно вместно этого удобней словами описать что ваш код умеет делать и добавить ссылку на документацию (которую можно автоматически генерировать по докстрингам), тогда вручную следить за расхождениями придётся сильно меньше... особенно если генерировать документацию в pre-commit hook.
 
2) Непоследовательно: одна часть имён `выделена`, другая - нет

File: README.md [hunk 3]
```diff
@@ -1 +1,49 @@
-# sequence_processing
\ No newline at end of file
+# sequence processing
+
+The complex of bioinformatics tools implemented for homework 18 in [Bioinformatic Insitute](https://bioinf.me/).
```
Comment (iam28th) [inline README.md:3]: стоит подчеркнуть, что это не одна домашка, а целый набор за год

File: README.md [hunk 4]
```diff
@@ -1 +1,49 @@
...
+## `bio_files_processor` module
+
+* `convert_multiline_fasta_to_oneline` - convert multiline fasta-file to oneline file.
+* `select_genes_from_gbk_to_fasta` - extract the chosen genes from gbk-file and write them to fasta-file
+* `OpenFasta` - context manager for opening and reading fasta-files.
+
+
+## `custom_random_forest` module 
+
+*`RandomForestClassifierCustom` - Custom implementation of the RandomForestClassifierCustom based on sklearn with threading.
```
Comment (iam28th) [inline README.md:44]: здесь пробела не хватает похоже

File: bio_files_processor.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+@dataclass
+class FastaRecord:
+    """DataClass for fasta-data with id, description, seq."""
+    id: str
+    description: str
+    seq: str
+
+    def __repr__(self):
+        return (f'id = {self.id}\ndescription = {self.description}\nsequence = {self.seq}\n')
```
Comment (iam28th) [inline bio_files_processor.py:159]: целиком печатать `self.seq` не здорово; если там в фасте лежит целая хромосома например, то repr получится не очень информативным
посмотрите, как это сделано в Biopython

File: bio_files_processor.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+class OpenFasta:
+    """
+    Context manager for opening and reading fasta-files.
+    methods:
+        - read_record: read fasta-record one by one; return FastaRecord.
+        - read_records: read whole fasta-file with all fasta-records; return list of FastaRecord.
+    """
+    def __init__(self, path):
+        self.path = path
```
Comment (iam28th) [inline bio_files_processor.py:170]: у path нужно проаннотировать тип

File: bio_files_processor.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        self.seq = seq
+        
+        if id_seq:
+            return FastaRecord(id=id_seq, description=descr, seq=seq)
+
+    def read_record(self)->FastaRecord:
+        """Read fasta-record one by one and return FastaRecord."""
+        return next(self)
+
+    def read_records(self)->list:
```
Comment (iam28th) [inline bio_files_processor.py:200]: у списков лучше дополнительно подписывать тип элемента
`-> list[FastaRecord]`

File: sequence_processing.py [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+        return str(self.seq)
+
+    def __str__(self):
+        return str(self.seq)
+
+    def is_alphabet_correct(self) -> bool:
+        """Check for compliance with the biological alphabet."""
+        return set(self.seq) <= set(self.alphabet)
+
+    def complement(self):
```
Comment (iam28th) [inline sequence_processing.py:73]: тут тоже нужна аннтоция

```python
from typing import Self
...
def complement(self) -> Self:
   ...
```

File: sequence_processing.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+        dir_name = os.path.join(os.path.dirname(input_path),'fastq_filtrator_results')
+        if not os.path.isdir(dir_name):
+            os.mkdir(dir_name)
+        output_filename = os.path.join(dir_name, output_filename)
+
+    if not output_filename[-6:] == '.fastq':
+        output_filename = output_filename + '.fastq'
+
+    if os.path.isfile(output_filename):
+        raise ValueError(f'File with name {output_filename} is exist. Please, use another name for your output file.')
```
Comment (iam28th) [inline sequence_processing.py:233]: не в тему дз, но лучше проверку делать вначале функции

представьте что кто-то пофильтровал огромную фасту, а потом функция упала...

File: test_sequence_processing.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    return inp
+
+def test_dna_is_child(input_data):
+    inp = DNASequence(input_data)
+    assert isinstance(inp, NucleicAcidSequence)    
+
+def test_dna_complement_error():
+    inp = DNASequence('ATGNMK')
+    with pytest.raises(NotImplementedError):
+        inp.complement()
```
Comment (iam28th) [inline test_sequence_processing.py:21]: `NotImplementedError` по смыслу для чего-то, что ещё не сделано, но когда-то в будущем возможно появится

Здесь явно нужна какая-то другая ошибка, `ValueError` например или кастомная

File: custom_random_forest.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    
+    def predict_proba(self, X):
+        y_pred_s = [self._predict_proba_loop(X, i) for i in range(self.n_estimators)]
+        return np.mean(y_pred_s, axis=0)
+
+    def predict(self, X):
+        probas = self.predict_proba(X)
+        predictions = np.argmax(probas, axis=1)
+
+        return predictions
```
Comment (iam28th) [inline custom_random_forest.py:54]: `predict` тоже нужно распараллелить, -10б
в остальном выглядит хорошо, можно разве что типизацию добавить

File: custom_random_forest.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
+import numpy as np
+import multiprocessing as mp
+from functools import partial
+from sklearn.base import BaseEstimator
+from sklearn.tree import DecisionTreeClassifier
```
Comment (iam28th) [inline custom_random_forest.py:5]:
```suggestion
import multiprocessing as mp
from functools import partial

import numpy as np
from sklearn.base import BaseEstimator
from sklearn.tree import DecisionTreeClassifier
```

File: sequence_processing.py [hunk 13]
```diff
@@ -0,0 +1,362 @@
+import os
+import sys
+import io
+import datetime
+import requests
+
+from dotenv import load_dotenv
+from dataclasses import dataclass
+from abc import ABC, abstractmethod
+from Bio import SeqIO, SeqUtils, SeqRecord
```
Comment (iam28th) [inline sequence_processing.py:10]:
```suggestion
import datetime
import io
import os
import sys
from abc import ABC, abstractmethod
from dataclasses import dataclass

import requests
from Bio import SeqIO, SeqRecord, SeqUtils
from dotenv import load_dotenv
```

File: test_sequence_processing.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
+import os
+import pytest
```
Comment (iam28th) [inline test_sequence_processing.py:2]:
```suggestion
import os

import pytest
```

File: test_sequence_processing.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    file_path = 'tmp.fastq'
+    yield file_path
+    if os.path.exists(file_path):
+        os.remove(file_path)
+
+def test_write_fastq_exists(tmp_file):#
+    input_path = '../data/example_data/test_fastq.fastq'
+    filter_fastq(input_path, tmp_file)
+    assert os.path.exists(tmp_file)
+
```
Comment (iam28th) [inline test_sequence_processing.py:62]: пустые строки в конце файла не нужны

File: test_sequence_processing.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    assert isinstance(result, target)
+
+def test_transcribed_dna(input_data):
+    inp = DNASequence(input_data)
+    target = 'AuGc'
+    result = str(DNASequence(inp).transcribe())   
+    assert target == result        
+
+def test_rna_alphabet(input_data):
+    target = {'A', 'C', 'G', 'U', 'a', 'c', 'g', 'u'}
```
Comment (iam28th) [inline test_sequence_processing.py:36]:
```python
target = set("ACGUacgu")
```

File: test_sequence_processing.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+def tmp_file():
+    file_path = 'tmp.fastq'
+    yield file_path
+    if os.path.exists(file_path):
+        os.remove(file_path)
+
+def test_write_fastq_exists(tmp_file):#
+    input_path = '../data/example_data/test_fastq.fastq'
+    filter_fastq(input_path, tmp_file)
+    assert os.path.exists(tmp_file)
```
Comment (iam28th) [inline test_sequence_processing.py:61]: тут можно было что-то более хитрое сделать, например пофильтровать по длине, затем вручную пройтись  и убедиться что длины лежат в диапазоне

**Review by iam28th — COMMENTED** [review]
Ольга, добрый день!

Оформление репозитория в целом хорошее, код понравился.
Разве что кое-где не достаёт type hint-ов, и порядок импортов не соответствует общепринятому - за это -2б.
Для автоматизации порядка импортов советую использовать isort или ruff.
Ещё было бы неплохо показать в ноутбуке содержимое файлов `example_fasta.fasta` и `example_multiline_fasta.fasta`:
`!cat example_fasta.fasta` (только тогда файлы поменьше сделать, чтобы не заполнили собой весь экран).
13/15

В примере с лесом:
- оба раза пишите "Время выполнения ... с 1 потоком";
- полное совпадение всех предсказанных значений с истинными - весьма маловероятное событие, поэтому такая проверка не очень информативна; лучше какую-нибудь метрику посчитать, например Accuracy;
- ещё какое-то время думал, почему в 2 потока аж в 100 раз быстрее стало...

В общем этот пример не очень аккуратный получился, -1 балл.
14/25

С тестами всё хорошо, 10/10

Итого у вас 37/50 баллов за это дз.
